### PR TITLE
Add option to track widget rebuilds and repaints from the Flutter inspector.

### PR DESCRIPTION
**Pull request by jacob314**

### CODE REVIEW

**Comment by jacob314** [comment]
When the bool `ext.flutter.inspector.trackRebuildDirtyWidgets` service extension is enabled,
dispatch `Flutter.RebuiltWidgets` events containing a compact representation of what source code
lines had widgets rebuilt and how many times they were rebuilt.
Similarly for repaints when `ext.flutter.inspector.trackRepaintWidgets` is true dispatch compact `Flutter.RepaintWidgets` describing repaints. The repaint case is a bit more experimental as the mapping from RenderObjects that were repainted and widgets is a bit more fuzzy as not all widgets have render objects so we need to bread the information up the widget tree for widgets that didn't directly have a render object.

Here is the existing UI in intellij that visualizes the rebuild counts.
cropped

We are also using the counts in Intellij to generate tips to suggest fixes for common performance problems users encounter. For example, if you have widgets inside a stateful widget that rebuild a lot, look at the article describing stateful widget best practices.

File: packages/flutter/lib/src/widgets/debug.dart [hunk 1]
```diff
@@ -30,6 +30,25 @@ import 'table.dart';
 /// See also the discussion at [WidgetsBinding.drawFrame].
 bool debugPrintRebuildDirtyWidgets = false;
 
+/// Signature for [debugOnRebuildDirtyWidget] implementations.
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/debug.dart]: @Hixie The most controversial part of the CL is adding the two debugging hooks
debugOnRebuildDirtyWidget and debugOnProfilePaint.
What I would like is a way for the inspector to be notified over each widget rebuild and render object repaint.
This seems like the most minimal option to support this use case in debug builds but I'm open to other options.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 2]
```diff
@@ -691,6 +692,12 @@ class _SerializeConfig {
     includeProperties = base.includeProperties,
     expandPropertyValues = base.expandPropertyValues;
 
+  /// Optional object group name used to manage manage lifetimes of object
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: Making groupName optional is added as part of this CL so I can get lightweight info on the Element tree without object ids for use by logic in IntelliJ to generate lint errors given rebuild counts and a snapshot of the widget tree to determine whether frequently built widgets had parents of specific types.

File: packages/flutter/lib/src/widgets/debug.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/debug.dart]: Seems fine to me, the way you added it is very clean. I actually thought the most controversial part was the addition of more hooks in widgets/binding.dart that defer to the WidgetInspectorService singleton. :-)

**Comment by Hixie** [comment]
cc @goderbauer for more thorough review.

File: packages/flutter/lib/src/widgets/debug.dart [hunk 3]
```diff
@@ -30,6 +30,25 @@ import 'table.dart';
...
+typedef RebuildDirtyWidgetCallback = void Function(Element e, bool builtOnce);
+
+/// Callback invoked for every dirty widget built each frame.
+///
+/// This callback is only invoked in debug builds.
+///
+/// See also:
+///  * [debugPrintRebuildDirtyWidgets], which does something similar but logs
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/flutter/lib/src/widgets/debug.dart]: Done. Fixed another place in the inspector I'd missed the \n as well.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 4]
```diff
@@ -691,6 +692,12 @@ class _SerializeConfig {
...
+  /// references in the returned JSON.
+  ///
+  /// A call to `ext.flutter.inspector.disposeGroup` is required before objects
+  /// in the tree are garbage collected unless [groupName] is `null` in
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/flutter/lib/src/widgets/widget_inspector.dart]: done. Stripped all `null`, `true`, and `false`.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 5]
```diff
@@ -928,6 +946,8 @@ mixin WidgetInspectorService {
     assert(!_debugServiceExtensionsRegistered);
     assert(() { _debugServiceExtensionsRegistered = true; return true; }());
 
+    WidgetsBinding.instance.addPersistentFrameCallback(_onFrameStart);
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/flutter/lib/src/widgets/widget_inspector.dart]: done

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 6]
```diff
@@ -940,6 +960,60 @@ mixin WidgetInspectorService {
...
+          if (value) {
+            assert(debugOnProfilePaint == null);
+            debugOnProfilePaint = _onPaint;
+            // Trigger an immediate paint so the user has some baseline painting
+            // stats to view.
+            void markTreeNeedsPaint(RenderObject renderObject) {
+              renderObject.markNeedsPaint();
+              renderObject.visitChildren(markTreeNeedsPaint);
+            }
+            final RenderObject root = WidgetsBinding.instance.renderViewElement?.renderObject;
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/flutter/lib/src/widgets/widget_inspector.dart]: done

File: packages/flutter/lib/src/widgets/debug.dart (hunk 1 above)
Comment (jacob314) [reply packages/flutter/lib/src/widgets/debug.dart]: thanks for the feedback!

File: packages/flutter/lib/src/rendering/debug.dart [hunk 7]
```diff
@@ -116,6 +118,23 @@ bool debugCheckIntrinsicSizes = false;
...
+typedef ProfilePaintCallback = void Function(RenderObject renderObject);
+
+/// Callback invoked for every [RenderObject] painted each frame.
+///
+/// This callback is only invoked in debug builds.
+///
+/// See also:
+///  * [debugProfilePaintsEnabled], which does something similar but adds
+///    [dart:developer.Timeline] events instead of invoking a callback.
+///  * [debugOnBuild], which does something similar for widgets being built.
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/flutter/lib/src/rendering/debug.dart]: done. Good catch. I'd tweaked the name back to 
debugOnRebuildDirtyWidget while iterating on the design and failed to update the comment.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 8]
```diff
@@ -1734,13 +1824,237 @@ mixin WidgetInspectorService {
...
+      _postStatsEvent('Flutter.RepaintWidgets', _repaintStats);
+    }
+  }
+
+  void _postStatsEvent(String eventName, _ElementLocationStatsTracker stats) {
+    postEvent(eventName, stats.exportToJson(_frameStart));
+  }
+
+  /// All events dispatched by a [WidgetInspectorService] use this method
+  /// instead of calling [developer.postEvent] directly  so that tests for
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/flutter/lib/src/widgets/widget_inspector.dart]: done

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 9]
```diff
@@ -1734,13 +1824,237 @@ mixin WidgetInspectorService {
...
+  final _ElementLocationStatsTracker _repaintStats = _ElementLocationStatsTracker();
+
+  void _onRebuildWidget(Element element, bool builtOnce) {
+    _rebuildStats.add(element);
+  }
+
+  void _onPaint(RenderObject renderObject) {
+    try {
+      final Element element = renderObject.debugCreator?.element;
+      assert(element is RenderObjectElement);
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/flutter/lib/src/widgets/widget_inspector.dart]: We should handle the null case. Done.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 10]
```diff
@@ -1734,13 +1824,237 @@ mixin WidgetInspectorService {
...
+      _repaintStats.add(element);
+
+      element.visitAncestorElements((Element ancestor) {
+        if (ancestor is RenderObjectElement) {
+          // This ancestor has its onw RenderObject. Don't conflate it.
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/flutter/lib/src/widgets/widget_inspector.dart]: Updated this comment added a higher level comment clarifying what I am trying to do here.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 11]
```diff
@@ -1734,13 +1824,237 @@ mixin WidgetInspectorService {
...
+  ///
+  /// Only locations local to the current project are included as a performance
+  /// optimization.
+  final List<_LocationCount> newLocations = <_LocationCount>[];
+
+  /// Increments the count associated with the creation location of [element] if
+  /// the creation location is local to the current project.
+  void add(Element element) {
+    final Object widget = element.widget;
+    if (widget is _HasCreationLocation) {
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/flutter/lib/src/widgets/widget_inspector.dart]: agreed that is nicer. Unfortunately Dart inferences isn't handling the is! then return case well so I had to add a manual cast.

File: packages/flutter/lib/src/rendering/debug.dart [hunk 12]
```diff
@@ -116,6 +118,23 @@ bool debugCheckIntrinsicSizes = false;
...
 
+/// Signature for [debugOnProfilePaint] implementations.
+typedef ProfilePaintCallback = void Function(RenderObject renderObject);
+
+/// Callback invoked for every [RenderObject] painted each frame.
+///
+/// This callback is only invoked in debug builds.
+///
+/// See also:
+///  * [debugProfilePaintsEnabled], which does something similar but adds
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/flutter/lib/src/rendering/debug.dart]: done.

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**